src/spin_mutex.cc

```diff
@@ -3,19 +3,40 @@
 #include <cstddef>
 #include <thread>
 
+#define _ASM_NOOP asm volatile("nop")
+
+// Cannot use `do { } while (0)` here because there would be huge differences between -O0 and -O3
+#define _2X(statement) \
+    {                  \
+        statement;     \
+        statement;     \
+    }
+
+#define _2X_NOOP _2X(_ASM_NOOP)
+#define _4X_NOOP _2X(_2X_NOOP)
+#define _8X_NOOP _2X(_4X_NOOP)
+#define _16X_NOOP _2X(_8X_NOOP)
+#define _32X_NOOP _2X(_16X_NOOP)
+
 namespace cris::core {
 
 void HybridSpinMutex::lock() {
-    // 50k loop ~= 500k x86 inst ~= 1M cycles ~= 500us@2GHz
-    // To reduce contention, 100 loop ~= 1k x86 inst ~= 2k cycles ~= 1us@2GHz between atomic variable checks.
-    constexpr std::size_t kSpinningLimit             = 50000;
-    constexpr std::size_t kSelfSpinningBetweenChecks = 100;
-
-    for (std::size_t i = 0; i < kSpinningLimit / kSelfSpinningBetweenChecks; ++i) {
+    // Spinning for about 500us@3GHz
```

Comment:
This doesn't sound very accurate.
Maybe you can check the clock every 10-100us.
IIRC now() takes around us scale, so this should be enough to make it negligible.

Reply:
No, we don't need to be super accurate in this case. If the freq goes up, then everything goes faster, and the critical section protected by spinlock should not do anything like "wait for 100ms" or I/O. So we don't care if 2x the frequency will make the time shorter. It does not matter for me.

Reply:
> No, we don't need to be super accurate in this case.

But we can.

> If the freq goes up, then everything goes faster, and the critical section protected by spinlock should not do anything like "wait for 100ms" or I/O.

Hybrid lock's spin time should be proportional to both ctx sw overhead and the time slice, not system speed.
ctx sw overhead is negligible comparing to scheduling jitter at ms scale.
So we should be targeting maybe 1-10% of a time slice.

My server says 100 in `cat /proc/sys/kernel/sched_rr_timeslice_ms` and `cat /proc/sys/kernel/sched_min_granularity_ns` from 3-10M, so 500us sounds reasonable to me.

Reply:
If you really think it is necessary, feel free to add it as a follow-up